### PR TITLE
Adding InputManager+ BarrelArranger

### PR DESCRIPTION
**Pull request by EmaSuriano**
# What's here? 

- Creation of MultiplexorHandler --> this class has the responsibility to communicate with the multiplexor, initializing the pins, applying delays, etc.
- Creation of InputManager --> this class has the responsibility of knowing from where to retrieve the input, in order to process it after.
- Creation of BarrelArranger --> this class has the responsibility to know what's the shape of a message to rotate a barrel, and also how to do it.
- A lot of code has been moved to this classes, so it looks a lot cleaner.

### CODE REVIEW

File: Arduino.ino [hunk 1]
```diff
@@ -8,62 +11,22 @@ int CLOCK_PIN = 10;     // Pin connected to SH_CP of 74HC595
...
-const int SERIAL_PORT = 1;
-const int MANUAL_MODE = 2;
-char *words[3] = {"AAAA", "CCCC", "BBBB"};
-
-WordRepresenter wordRepresenter(LATCH_PIN, CLOCK_PIN, DATA_PIN, DELAY_BETWEEN_MOTOR_STEPS);
-int selectedMode = SERIAL_PORT;
+MultiplexorHandler multiplexorHandler(LATCH_PIN, CLOCK_PIN, DATA_PIN);
+WordRepresenter wordRepresenter(multiplexorHandler);
+BarrelArranger barrelArranger(multiplexorHandler, DIRECTION_PIN);
+InputManager inputManager(1);
```
Comment (santicanti) [inline Arduino.ino:15]: Que es el 1?

File: BarrelArranger.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+bool BarrelArranger::hasToMoveBarrel(char *word)
+{
+    return word[0] == MOVE_BARREL_CHARACTER;
+}
+
+// expected shape of word when moving a barrel --> -D1 (D is the corresponding direction and 1 is the motor number)
+void BarrelArranger::moveBarrel(char *word)
+{
+    int direction = word[1] == 'D' ? RIGHT_DIRECTION : LEFT_DIRECTION;
+    int motorNumber = (int)word[2] - 48;
```
Comment (santicanti) [inline BarrelArranger.cpp]: magic number

File: BarrelArranger.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+void BarrelArranger::moveMotor(int motorNumber, int direction)
+{
+    if (_lastDirection != direction)
+    {
+        digitalWrite(_directionPin, direction);
+    }
+
+    _lastDirection = direction;
```
Comment (santicanti) [inline BarrelArranger.cpp]: Adentro del if de arriba

File: BarrelArranger.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+{
+    int direction = word[1] == 'D' ? RIGHT_DIRECTION : LEFT_DIRECTION;
+    int motorNumber = (int)word[2] - 48;
+
+    moveMotor(motorNumber, direction);
+}
+
+void BarrelArranger::moveMotor(int motorNumber, int direction)
+{
+    if (_lastDirection != direction)
```
Comment (santicanti) [inline BarrelArranger.cpp:82]: Agregar comentario de esto

File: BarrelArranger.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    _lastDirection = direction;
+
+    for (int side = 0; side < STEPS_PER_MOVE; side++)
+    {
+        _multiplexorHandler.sendMultiplexorData(MULTIPLEXOR_DATA_BY_MOTOR_NUMBER[motorNumber]);
+    }
+
+    if (direction == LEFT_DIRECTION)
```
Comment (santicanti) [inline BarrelArranger.cpp]: Agregar comentario explicando esto

File: InputManager.cpp [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        Serial.readString().toCharArray(result, 5);
+    }
+    return result;
+}
+
+char *InputManager::getHardcodedWord()
+{
+    char *words[3] = {"AAAA", "CCCC", "BBBB"};
+    delay(DELAY_BETWEEN_WORDS);
+    return words[_hardcodedWordIndex++];
```
Comment (santicanti) [inline InputManager.cpp]: return words[_hardcodedWordIndex++ % 3];  sino se va del array

File: MultiplexorHandler.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+void MultiplexorHandler::sendMultiplexorData(int data)
+{
+    digitalWrite(_latchPin, LOW);
+    shiftOut(_dataPin, _clockPin, LSBFIRST, data);
+    digitalWrite(_latchPin, HIGH);
+    delay(DELAY_BETWEEN_MOTOR_STEPS);
+
+    digitalWrite(_latchPin, LOW);
+    shiftOut(_dataPin, _clockPin, LSBFIRST, 0);
```
Comment (santicanti) [inline MultiplexorHandler.cpp]: Extraer 0 a constante NO_DATA

**Review by santicanti — CHANGES_REQUESTED** [review]

File: Arduino.ino (hunk 1 above)
Comment (EmaSuriano) [reply Arduino.ino:15]: Es el modo, que esta declarado en el otro lado. No se me ocurrio otra forma de instanciarlo :/

File: BarrelArranger.cpp (hunk 4 above)
Comment (EmaSuriano) [reply BarrelArranger.cpp:82]: Erm para mi no hace falta, se entiende.

File: Arduino.ino (hunk 1 above)
Comment (santicanti) [reply Arduino.ino:15]: Pero qué es? Jajaja

**Pull request closed without merging**